File: projects/ProtoMine/ProtoMine.Core/Protocol/Packets/HandshakePacket.cs
using ProtoMine.Core.Protocol.Packets.Base;
using ProtoMine.Core.Protocol.Types;

namespace ProtoMine.Core.Protocol.Packets;

public class HandshakePacket : Packet
{
	public override int PacketId => 0x00;

	public HandshakePacket(
		int protocolVersion,
		string serverAddress,
		ushort serverPort,
		int nextState
	)
	{
		NextState = nextState;
		ServerPort = serverPort;
		ServerAddress = serverAddress;
		ProtocolVersion = protocolVersion;
	}

	[ProtocolProperty(0, typeof(ProtocolInt))]
	public int ProtocolVersion { get; }

	[ProtocolProperty(1, typeof(ProtocolString))]
	public string ServerAddress { get; }

	[ProtocolProperty(2, typeof(ProtocolUShort))]
	public ushort ServerPort { get; }

	[ProtocolProperty(3, typeof(ProtocolVarInt))]
	public int NextState { get; }
}

File: projects/MinePack/lib/protocol/packets/v761/HandshakePacket.ts
import { Packet } from "../Packet.js";
import { PacketProperty } from "../PacketProperty.js";
import { ProtocolContext } from "../../ProtocolContext.js";

import {
	ProtocolState,
	ProtocolVersion,
	SendablePacket,
} from "../VersionedPacket.js";

import * as types from "../../../core/types/index.js";

@SendablePacket(ProtocolVersion.v761, ProtocolState.Handshake)
export class HandshakePacket extends Packet {
	public readonly packetId = 0x00;

	@PacketProperty(1, types.VarInt)
	public protocolVersion!: number;

	@PacketProperty(2, types.VarString)
	public serverAddress!: string;

	@PacketProperty(3, types.UnsignedShort)
	public serverPort!: number;

	@PacketProperty(4, types.VarInt)
	public nextState!: number;

	public updateProtocolContext(context: ProtocolContext) {
		super.updateProtocolContext(context);

		context.state.update(
			this.nextState === 1 ? ProtocolState.Status : ProtocolState.Login
		);
	}
}
